File: pawpause/pomodoro.py
import threading
from enum import Enum


class Phase(Enum):
    WORK = "work"
    SHORT_BREAK = "short_break"
    LONG_BREAK = "long_break"


class RunState(Enum):
    IDLE = "idle"
    RUNNING = "running"
    PAUSED = "paused"


PHASE_LABELS = {
    Phase.WORK: "Work",
    Phase.SHORT_BREAK: "Short break",
    Phase.LONG_BREAK: "Long break",
}

BREAK_PHASES = (Phase.SHORT_BREAK, Phase.LONG_BREAK)


class Pomodoro:
    """A minimal Pomodoro state machine, driven by an external 1s ticker.

    on_transition(old_phase, new_phase) fires on every phase change,
    including start (None -> WORK) and stop (phase -> None).
    on_tick() fires once per call to tick(), for UI refresh.
    """

    def __init__(self, config, on_transition, on_tick):
        self.config = config
        self.on_transition = on_transition
        self.on_tick = on_tick

        self.state = RunState.IDLE
        self.phase = None
        self.remaining = 0
        self.session_count = 0
        self._lock = threading.Lock()

    def _duration(self, phase):
        minutes = {
            Phase.WORK: self.config.work_minutes,
            Phase.SHORT_BREAK: self.config.short_break_minutes,
            Phase.LONG_BREAK: self.config.long_break_minutes,
        }[phase]
        return int(minutes * 60)

    def _advance(self, phase, session_count):
        if phase == Phase.WORK:
            session_count += 1
            if session_count % self.config.sessions_before_long_break == 0:
                return Phase.LONG_BREAK, session_count
            return Phase.SHORT_BREAK, session_count
        return Phase.WORK, session_count

    def start(self):
        with self._lock:
            if self.state != RunState.IDLE:
                return
            self.phase = Phase.WORK
            self.remaining = self._duration(Phase.WORK)
            self.session_count = 0
            self.state = RunState.RUNNING
        self.on_transition(None, Phase.WORK)

    def pause(self):
        with self._lock:
            if self.state == RunState.RUNNING:
                self.state = RunState.PAUSED

    def resume(self):
        with self._lock:
            if self.state == RunState.PAUSED:
                self.state = RunState.RUNNING

    def skip(self):
        with self._lock:
            if self.phase is None:
                return
            old_phase = self.phase
            new_phase, self.session_count = self._advance(old_phase, self.session_count)
            self.phase = new_phase
            self.remaining = self._duration(new_phase)
        self.on_transition(old_phase, new_phase)

    def stop(self):
        with self._lock:
            old_phase = self.phase
            self.phase = None
            self.state = RunState.IDLE
            self.remaining = 0
            self.session_count = 0
        if old_phase is not None:
            self.on_transition(old_phase, None)

    def tick(self):
        """Call once per second from a background thread."""
        transition = None
        with self._lock:
            if self.state == RunState.RUNNING and self.phase is not None:
                self.remaining -= 1
                if self.remaining <= 0:
                    old_phase = self.phase
                    new_phase, self.session_count = self._advance(old_phase, self.session_count)
                    self.phase = new_phase
                    self.remaining = self._duration(new_phase)
                    transition = (old_phase, new_phase)
        if transition:
            self.on_transition(*transition)
        self.on_tick()

    def status_text(self):
        with self._lock:
            if self.phase is None:
                return "Idle"
            mm, ss = divmod(max(self.remaining, 0), 60)
            label = PHASE_LABELS[self.phase]
            suffix = " (paused)" if self.state == RunState.PAUSED else ""
            return f"{label} — {mm:02d}:{ss:02d} left{suffix}"

File: pawpause-applet/src/pomodoro.rs
use crate::config::Config;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Phase {
    Work,
    ShortBreak,
    LongBreak,
}

impl Phase {
    pub fn label(self) -> &'static str {
        match self {
            Phase::Work => "Work",
            Phase::ShortBreak => "Short break",
            Phase::LongBreak => "Long break",
        }
    }

    pub fn is_break(self) -> bool {
        matches!(self, Phase::ShortBreak | Phase::LongBreak)
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum RunState {
    Idle,
    Running,
    Paused,
}

/// A phase transition, with how long the phase being left had actually run
/// for (natural completion, skip, or stop all count) — used to credit focus
/// time to the right task/project.
pub struct Transition {
    pub old_phase: Option<Phase>,
    pub new_phase: Option<Phase>,
    pub old_phase_elapsed_secs: i64,
}

/// A minimal Pomodoro state machine, driven by an external 1s tick().
pub struct Pomodoro {
    pub state: RunState,
    pub phase: Option<Phase>,
    pub remaining: i64,
    pub session_count: u32,
    /// The full planned duration of the current phase, so elapsed time can be
    /// recovered as `current_phase_duration - remaining` when it ends.
    current_phase_duration: i64,
}

impl Pomodoro {
    pub fn new() -> Self {
        Self {
            state: RunState::Idle,
            phase: None,
            remaining: 0,
            session_count: 0,
            current_phase_duration: 0,
        }
    }

    fn elapsed_in_current_phase(&self) -> i64 {
        (self.current_phase_duration - self.remaining.max(0)).max(0)
    }

    fn duration(phase: Phase, config: &Config) -> i64 {
        let minutes = match phase {
            Phase::Work => config.work_minutes,
            Phase::ShortBreak => config.short_break_minutes,
            Phase::LongBreak => config.long_break_minutes,
        };
        (minutes * 60.0).round() as i64
    }

    fn advance(phase: Phase, session_count: u32, config: &Config) -> (Phase, u32) {
        match phase {
            Phase::Work => {
                let session_count = session_count + 1;
                if config.sessions_before_long_break > 0
                    && session_count % config.sessions_before_long_break == 0
                {
                    (Phase::LongBreak, session_count)
                } else {
                    (Phase::ShortBreak, session_count)
                }
            }
            Phase::ShortBreak | Phase::LongBreak => (Phase::Work, session_count),
        }
    }

    /// Returns Some(transition) if a transition occurred.
    pub fn start(&mut self, config: &Config) -> Option<Transition> {
        if self.state != RunState::Idle {
            return None;
        }
        self.phase = Some(Phase::Work);
        self.current_phase_duration = Self::duration(Phase::Work, config);
        self.remaining = self.current_phase_duration;
        self.session_count = 0;
        self.state = RunState::Running;
        Some(Transition {
            old_phase: None,
            new_phase: self.phase,
            old_phase_elapsed_secs: 0,
        })
    }

    pub fn pause(&mut self) {
        if self.state == RunState::Running {
            self.state = RunState::Paused;
        }
    }

    pub fn resume(&mut self) {
        if self.state == RunState::Paused {
            self.state = RunState::Running;
        }
    }

    pub fn skip(&mut self, config: &Config) -> Option<Transition> {
        let old_phase = self.phase?;
        let old_phase_elapsed_secs = self.elapsed_in_current_phase();
        let (new_phase, session_count) = Self::advance(old_phase, self.session_count, config);
        self.phase = Some(new_phase);
        self.session_count = session_count;
        self.current_phase_duration = Self::duration(new_phase, config);
        self.remaining = self.current_phase_duration;
        Some(Transition {
            old_phase: Some(old_phase),
            new_phase: Some(new_phase),
            old_phase_elapsed_secs,
        })
    }

    pub fn stop(&mut self) -> Option<Transition> {
        let old_phase = self.phase;
        let old_phase_elapsed_secs = self.elapsed_in_current_phase();
        self.phase = None;
        self.state = RunState::Idle;
        self.remaining = 0;
        self.current_phase_duration = 0;
        self.session_count = 0;
        old_phase.map(|p| Transition {
            old_phase: Some(p),
            new_phase: None,
            old_phase_elapsed_secs,
        })
    }

    /// Call once per second. Returns Some(transition) if a phase auto-advanced
    /// this tick.
    pub fn tick(&mut self, config: &Config) -> Option<Transition> {
        if self.state != RunState::Running {
            return None;
        }
        let Some(phase) = self.phase else {
            return None;
        };
        self.remaining -= 1;
        if self.remaining > 0 {
            return None;
        }
        let old_phase_elapsed_secs = self.current_phase_duration;
        let (new_phase, session_count) = Self::advance(phase, self.session_count, config);
        self.phase = Some(new_phase);
        self.session_count = session_count;
        self.current_phase_duration = Self::duration(new_phase, config);
        self.remaining = self.current_phase_duration;
        Some(Transition {
            old_phase: Some(phase),
            new_phase: Some(new_phase),
            old_phase_elapsed_secs,
        })
    }

    pub fn status_text(&self) -> String {
        match self.phase {
            None => "Idle".to_string(),
            Some(phase) => {
                let remaining = self.remaining.max(0);
                let mm = remaining / 60;
                let ss = remaining % 60;
                let suffix = if self.state == RunState::Paused {
                    " (paused)"
                } else {
                    ""
                };
                format!("{} — {mm:02}:{ss:02} left{suffix}", phase.label())
            }
        }
    }

    pub fn short_time_text(&self) -> String {
        match self.phase {
            None => "—".to_string(),
            Some(_) => {
                let remaining = self.remaining.max(0);
                format!("{:02}:{:02}", remaining / 60, remaining % 60)
            }
        }
    }
}
